Fix wordbreak for URLs

jasmussen opened the pull request:
## Description

Fixes #36961. URLs didn't wrap:

<img width="787" alt="before" src="https://user-images.githubusercontent.com/1204802/144026109-7ad1d374-88a6-4ac2-8fb7-2f23c6abdc17.png">

They do after this PR:

<img width="715" alt="Screenshot 2021-11-30 at 10 58 46" src="https://user-images.githubusercontent.com/1204802/144026145-342862ff-c8a9-4106-9a19-96d6140993c4.png">

For what it's worth, I don't personally think it's that valuable to see the full URL like this, especially when it gets clipped in editing it after the fact:

<img width="672" alt="Screenshot 2021-11-30 at 10 58 53" src="https://user-images.githubusercontent.com/1204802/144026244-d6a97841-fb14-454f-8cdf-ada0a76e37fb.png">

But that's a not at all a strong opinion, so I've kept the behavior as is, given the CSS comment pointed to a particular issue.

## How has this been tested?

Add a very long URL that does not include hyphens, and observe that it still gets broken up.

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Fix wordbreak for URLs

getdave commented:
This makes me happy! On my review list for today 🙇

Mamaduka commented:
We also have PR with an alternative fix - #36980.

jasmussen commented:
I don't have a strong opinion on which fix is worth landing, but for some context on the additional properties I added, [CSS tricks has a great rundown](https://css-tricks.com/where-lines-break-is-complicated-heres-all-the-related-css-and-html/#a-subtle-fix-overflow-wrap-break-word).

Mamaduka commented:
Thanks, @jasmussen.

I also don't have a strong opinion on this.

ntsekouras commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -197,12 +197,14 @@ $preview-image-height: 140px;
 		white-space: pre-wrap;
+		overflow-wrap: break-word;
 		word-wrap: break-word;
+		word-break: break-word;
```
Do we need this? It's [deprecated](https://developer.mozilla.org/en-US/docs/Web/CSS/word-break)

jasmussen replied:
We do need `word-break`, yes, that's what fixed the issue. But good call, I moved to `break-all` instead of `break-word` 👍 👍

<img width="901" alt="Screenshot 2021-11-30 at 12 20 29" src="https://user-images.githubusercontent.com/1204802/144038455-aad8476e-85d6-4d09-8162-1678b3ceca3c.png">

jasmussen committed:
Use break-all instead.

ntsekouras commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -197,12 +197,14 @@ $preview-image-height: 140px;
 		// of text, particular those with no spaces.
 		// See: https://github.com/WordPress/gutenberg/issues/33586#issuecomment-888921188
 		white-space: pre-wrap;
+		overflow-wrap: break-word;
```
Not related to this PR, but it seems that `word-wrap` has been renamed to [overflow-wrap](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-wrap). 

>The property was originally a nonstandard and unprefixed Microsoft extension called word-wrap, and was implemented by most browsers with the same name. It has since been renamed to overflow-wrap, with word-wrap being an alias.

I guess we don't need both properties here and in other places we have this prop.

jasmussen replied:
Confirmed, looks like it's the same without it.

ntsekouras replied:
I think we should remove the old one `word-wrap` and keep `overflow-wrap`, but it's a minor thing since it's an alias. Maybe we could replace all usages of `word-wrap` in another PR..

jasmussen replied:
I need to step away for a bit of a late lunch now. But feel free to push any changes to the PR. From my investigations, it looks like the key property that solved the problem was `word-break: break-all;`.

ntsekouras reviewed (approved):
Looks good Joen, thank! 👍 

Personally I think this approach is better because it fixes the issue with just one css rule that  does just what we want.

jasmussen committed:
Remove overflow-wrap.

walbo reviewed (approved):
No strong options about which approach is best. The difference between the PR's is that #36980 truncates the URL.

ntsekouras committed:
prefer the new css prop `overflow-wrap`

jasmussen commented:
I can't quite figure out why the PHP unit tests keep failing.

javierarce commented:
Ok, here's a "moderately warm" opinion 😅: given that users will still be able to see and edit the full URL, I'd prefer truncating the string as [the other PR](https://github.com/WordPress/gutenberg/pull/36980) does, because the popover will be generally shorter.  

Here's an example of the same URL using this PR (left) and the other one (right).

![image](https://user-images.githubusercontent.com/4933/144057865-028cfcbf-3915-4d45-b05d-50855470be4c.png)

Also, notice that this PR does something weird with the title.

getdave commented:
> I'd prefer truncating the string 

Please can I sound a slight warning bell here that I've had lots of feedback from users in the past that they often need to be able to see the full URL.

If we do go down this route for UX/UI reasons, please can we include a method for users to be able to see the full URL? Perhaps an expose on hover or a tooltip or something? 🙏 🙏 🙏 🙏 🙏

getdave reviewed (approved):
Thank you for picking this one up 🙇 

I have also tested and I like this PR because it involves no truncation of strings. The [property used has good x-browser support](https://caniuse.com/?search=overflow-wrap) as well.

As mentioned in https://github.com/WordPress/gutenberg/pull/36993#issuecomment-982665722 I recommend we don't truncate unless we provide a means by which the user can still access the full URL.

<img width="604" alt="Screenshot 2021-11-30 at 14 09 00" src="https://user-images.githubusercontent.com/444434/144062388-07d6686d-6418-4add-a06c-34d8398091c8.png">

javierarce commented:
> If we do go down this route for UX/UI reasons, please can we include a method for users to be able to see the full URL? Perhaps an expose on hover or a tooltip or something? 🙏 🙏 🙏 🙏 🙏

+1 this makes sense. We could also increase the number of the truncated characters so we showed two lines at most instead of always one.

ntsekouras commented:
>Also, notice that this PR does something weird with the title.

Oh.. this is valid. We should apply the css rule in the url only.

getdave commented:
Here is the PR where I added word wrapping instead of clipping originally in response to the user feedback

https://github.com/WordPress/gutenberg/pull/33753

jasmussen commented:
Thanks for the reviews. I'll take a look at the title when I get a moment!

jasmussen committed:
Only apply to URL.

jasmussen commented:
Okay, I rebased, and moved the `break-all` property to apply only to the URL. The title should also still wrap, but be broken per word rather than per letter. Before:
<img width="524" alt="Screenshot 2021-12-01 at 08 30 23" src="https://user-images.githubusercontent.com/1204802/144190953-b96a1b1e-f491-4940-bd45-67b245399fc6.png">

After:

<img width="520" alt="Screenshot 2021-12-01 at 08 31 32" src="https://user-images.githubusercontent.com/1204802/144190962-32871714-70ab-47f6-b13c-d20194c973f0.png">

Feel free to give this branch a spin, otherwise since I've addressed the feedback, I'll land it some time after the tests hopefully pass.

As with all software, we can always come back to this. Nothing is ever so perfect that it can't be improved upon.

getdave commented:
Thank you all for fixing this 🥳
